### PR TITLE
HPCC-17295 Add nested dataset support to R

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
Added new test to exercise this functionality.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@AttilaVamos I am building using the current version of Rcpp - 3.3 and RInside. Perhaps the build box should be updated?

@Michael-Gardner Should we include libRinside.so with the package for the R plugin to avoid version issues there (Rcpp is header only so we should be safe on that)

**Comment by AttilaVamos** [comment]
@richardkchapman 
On the smoketet machine there is:
```
Installed Packages
Name        : R
Arch        : x86_64
Version     : 3.3.1
Release     : 2.el7
```
and 
```
 RInside_0.2.13
```

**Comment by richardkchapman** [comment]
What version of Rcpp?

**Comment by AttilaVamos** [comment]
@richardkchapman 
```
Installed Packages
Name        : R-Rcpp
Arch        : x86_64
Version     : 0.12.9
Release     : 1.el7

Name        : R-Rcpp-devel
Arch        : x86_64
Version     : 0.12.9
Release     : 1.el7

Name        : R-Rcpp-examples
Arch        : x86_64
Version     : 0.12.9
Release     : 1.el7
```

**Comment by richardkchapman** [comment]
@ghalliday Please review/merge

**Comment by ghalliday** [comment]
@richardkchapman should the //nothor be removed?

File: plugins/Rembed/Rembed.cpp [hunk 1]
```diff
@@ -241,228 +241,340 @@ MODULE_EXIT()
...
-    virtual void processDecimal(const void *value, unsigned digits, unsigned precision, const RtlFieldInfo * field)
+    virtual void processDecimal(const void *value, unsigned digits, unsigned precision, const RtlFieldInfo * field) override
     {
         Decimal val;
         val.setDecimal(digits, precision, value);
-        Rcpp::List column = list[colIdx];
-        column[rowIdx] = val.getReal();
-        colIdx++;
+        if (inSet)
+            theRealSet[setIndex++] = (unsigned long int) value;
```
Comment (ghalliday) [inline plugins/Rembed/Rembed.cpp]: val.getReal()

File: plugins/Rembed/Rembed.cpp [hunk 2]
```diff
@@ -574,38 +790,29 @@ static size32_t getRowResult(RInside::Proxy &result, ARowBuilder &builder)
...
-            return NULL;
-        }
-        RtlDynamicRowBuilder builder(resultAllocator);
-        const RtlTypeInfo *typeInfo = builder.queryAllocator()->queryOutputMeta()->queryTypeInfo();
-        assertex(typeInfo);
-        RtlFieldStrInfo dummyField("<row>", NULL, typeInfo);
-        myRRowBuilder.setRowIdx(idx);
         try
         {
+            RtlFieldStrInfo dummyField("<row>", NULL, typeInfo);
```
Comment (ghalliday) [inline plugins/Rembed/Rembed.cpp]: I think you moved this to a member, but didn't delete this instance.

**Review by ghalliday — DISMISSED** [review]
@richardkchapman two comments

**Comment by Michael-Gardner** [comment]
@richardkchapman I'll look into including the Rcpp header and the RInside.so with the Rembed package.  They're GPL2/3 so they wouldn't be compatible shipping with the platform package, just the Rembed Plugin package.

**Comment by richardkchapman** [comment]
We don't need to ship the Rcpp header - just make sure that we use a modern enough one on our build machines. As coded it should be ok with 0.12.9 or 0.12.10 - not tested any earlier.

My concern was more with the RInside library - I don't think there's any way we can guarantee that the version of RInside lib that we pick up matches the version of the RInside header files we compiled against unless we include libRinside.so with our package, and experience of how often breaking changes are made in the RInside/Rcpp code I think it's likely to bite us sooner rather than later.

**Comment by richardkchapman** [comment]
@ghalliday Pushed changes following review. Let me know if happy and I will squash

**Comment by ghalliday** [comment]
@richardkchapman looks good, please squash

**Comment by Michael-Gardner** [comment]
@richardkchapman Done with the change to add RInside to the final package, and also added in a version check for rcpp (requires 0.12.1 or greater).  And I've updated the build boxes.  Do you want me to submit a PR to your branch or open a separate Jira ticket for the changes?

**Comment by richardkchapman** [comment]
@Michael-Gardner Separate is fine

**Comment by richardkchapman** [comment]
@ghalliday Please merge

**Pull request merged**